apollo-router/src/services/subgraph_service.rs

```diff
@@ -597,104 +593,205 @@ async fn call_http(
+                    graphql::Response::builder()
+                        .error(
+                            FetchError::SubrequestMalformedResponse {
+                                service: service_name.to_string(),
+                                reason: error.to_string(),
+                            }
+                            .to_graphql_error(None),
+                        )
+                        .build()
+                })
+            })
+        }
+        (ContentType::Other, _, Ok(_bytes)) => {
+            unreachable!(
+                "Should not have gotten here. We should not parse the body of a non json response"
+            )
+        }
+        (_, _, Err(err)) => {
+            // This will add any error that occurred during the fetch to the response including invalid content-type
+            graphql::Response::builder()
+                .error(err.to_graphql_error(None))
+                .build()
+        }
+    };
 
-    if display_body {
-        tracing::info!(
-            http.response.body = %String::from_utf8_lossy(&body), apollo.subgraph.name = %cloned_service_name, "Raw response body from subgraph {cloned_service_name:?} received"
-        );
+    // Add an error for response codes that are not 2xx
+    if !parts.status.is_success() {
+        let status = parts.status;
+        graphql.errors.push(
+            FetchError::SubrequestHttpError {
+                service: service_name.to_string(),
+                status_code: Some(status.as_u16()),
+                reason: format!(
+                    "{}: {}",
+                    status.as_str(),
+                    status.canonical_reason().unwrap_or("Unknown")
+                ),
+            }
+            .to_graphql_error(None),
+        )
     }
 
-    let graphql: graphql::Response =
-        tracing::debug_span!("parse_subgraph_response").in_scope(|| {
-            graphql::Response::from_bytes(&cloned_service_name, body).map_err(|error| {
-                FetchError::SubrequestMalformedResponse {
-                    service: cloned_service_name.clone(),
-                    reason: error.to_string(),
-                }
+    let resp = http::Response::from_parts(parts, graphql);
+    Ok(SubgraphResponse::new_from_response(resp, context))
+}
+
+async fn perform_fetch(
+    mut client: Decompression<Client<HttpsConnector<HttpConnector>>>,
+    service_name: &str,
+    request: Request<Body>,
+) -> Result<(Parts, Result<Bytes, FetchError>), FetchError> {
+    let response = match client.call(request).await {
+        Err(err) => {
+            tracing::error!(fetch_error = format!("{err:?}").as_str());
+            return Err(FetchError::SubrequestHttpError {
+                status_code: None,
+                service: service_name.to_string(),
+                reason: err.to_string(),
+            });
+        }
+        Ok(response) => response,
+    };
```

Comment:
Can't we use a `map_err` ?

Reply:
[7534cc6](https://github.com/apollographql/router/pull/3328/commits/7534cc6607c769c903bd62590d9d5a0c2c338b4f)